Add utility to get use_count of the TensorImpl a python tensor refers to

mikaylagawarecki opened the pull request:
Mirroring `torch._C._storage_Use_Count`

https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/csrc/cuda/Module.cpp#L1139-L1142

The way I've implemented the above PR, we need this to check the `use_count` of the tensor to see whether the `swap_tensors` path is valid, otherwise we fall back to setting `.data`. If there is another utility that lets me access `use_count` of `TensorImpl` from python I am happy to close this and use that instead

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #117167
* __->__ #117166
* #116955

mikaylagawarecki committed:
Add utility to get use_count of the TensorImpl a python tensor refers to

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add utility to get use_count of the TensorImpl a python tensor refers to"

Mirroring `torch._C._storage_Use_Count`

https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/csrc/cuda/Module.cpp#L1139-L1142

The way I've implemented the above PR, we need this to check the `use_count` of the tensor to see whether the `swap_tensors` path is valid, otherwise we fall back to setting `.data`. If there is another utility that lets me access `use_count` of `TensorImpl` from python I am happy to close this and use that instead




[ghstack-poisoned]

mikaylagawarecki commented on test/test_torch.py:
```diff
@@ -5658,6 +5658,16 @@ def test_masked_scatter_inplace_noncontiguous(self, device):
         actual = t_non_contig.masked_scatter_(mask_contig, source)
         self.assertEqual(actual, expected)
 
+    @unittest.skipIf(TEST_WITH_TORCHDYNAMO, "Dynamo doesn't init AccumulateGrad nodes on forward")
```
Does it make sense that Dynamo would not be doing this?

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -331,6 +331,16 @@ PyObject* THPModule_setDefaultDtype(PyObject* _unused, PyObject* dtype) {
+  HANDLE_TH_ERRORS
+  auto tensor_impl_ptr = THPUtils_unpackLong(cdata);
+  // NOLINTNEXTLINE(performance-no-int-to-ptr)
+  auto tensor_impl = (c10::TensorImpl*)tensor_impl_ptr;
```
Huh, this is not a raw pointer to a TensorImpl! It's an intrusive_ptr to a TensorImpl! There are more fields inlined!

mikaylagawarecki committed:
Update on "Add utility to get use_count of the TensorImpl a python tensor refers to"

Mirroring `torch._C._storage_Use_Count`

https://github.com/pytorch/pytorch/blob/7e37f63e5e77a217a08d9261af49e63ddb51d306/torch/csrc/cuda/Module.cpp#L1139-L1142

The way I've implemented the above PR, we need this to check the `use_count` of the tensor to see whether the `swap_tensors` path is valid, otherwise we fall back to setting `.data`. If there is another utility that lets me access `use_count` of `TensorImpl` from python I am happy to close this and use that instead




[ghstack-poisoned]

mikaylagawarecki commented:
No longer necessary as I chose not to do checks on parameters before using `swap_tensors` in `_apply()`
